syntax = "proto3";

package dragongem.trace;

message TraceEvent {
  message DynamicInst {
    message Memory { uint64 address = 1; }

    uint64 inst_id = 1;
    oneof inst_oneof {
      Memory memory = 3;
    }
  }
  
  message BBEnter {
    uint64 bb_id = 1;
  }

  message MissedInstCount {
    uint64 num_insts = 1;
  }

  message CallStack {
    message CallFrame {
      uint64 bb_id = 1;
      uint64 bb_exec_count = 2; // Number of retired instructions.
    }
    repeated CallFrame frames = 1; // From outer -> inner.
  }

  oneof event_oneof {
    DynamicInst inst = 1;
    BBEnter bb = 2;
    MissedInstCount missed = 3;
    CallStack call_stack = 4;
  }
}
